File: lib/ts/src/internal/schema/CatchHeaderCollisions.ts
import { TelepactSchemaParseError } from '../../TelepactSchemaParseError';
import { SchemaParseFailure } from '../../internal/schema/SchemaParseFailure';
import { getPathDocumentCoordinatesPseudoJson } from '../../internal/schema/GetPathDocumentCoordinatesPseudoJson';

export function catchHeaderCollisions(
    telepactSchemaNameToPseudoJson: Record<string, any[]>,
    headerKeys: Set<string>,
    keysToIndex: Record<string, number>,
    schemaKeysToDocumentNames: Record<string, string>,
    documentNamesToJson: Record<string, string>,
): void {
    const parseFailures: SchemaParseFailure[] = [];

    const headerKeysList = [...headerKeys];

    headerKeysList.sort((k1, k2) => {
        const documentName1 = schemaKeysToDocumentNames[k1];
        const documentName2 = schemaKeysToDocumentNames[k2];
        if (documentName1 !== documentName2) {
            return documentName1.localeCompare(documentName2);
        } else {
            const index1 = keysToIndex[k1];
            const index2 = keysToIndex[k2];
            return index1 - index2;
        }
    });

    for (let i = 0; i < headerKeysList.length; i++) {
        for (let j = i + 1; j < headerKeysList.length; j++) {
            const defKey = headerKeysList[i];
            const otherDefKey = headerKeysList[j];

            const index = keysToIndex[defKey];
            const otherIndex = keysToIndex[otherDefKey];

            const documentName = schemaKeysToDocumentNames[defKey];
            const otherDocumentName = schemaKeysToDocumentNames[otherDefKey];

            const telepactSchemaPseudoJson = telepactSchemaNameToPseudoJson[documentName];
            const otherTelepactSchemaPseudoJson = telepactSchemaNameToPseudoJson[otherDocumentName];

            const def = telepactSchemaPseudoJson[index] as Record<string, object>;
            const otherDef = otherTelepactSchemaPseudoJson[otherIndex] as Record<string, object>;

            const headerDef = def[defKey] as Record<string, object>;
            const otherHeaderDef = otherDef[otherDefKey] as Record<string, object>;

            const headerCollisions = Object.keys(headerDef).filter((k) => k in otherHeaderDef);
            for (const headerCollision of headerCollisions) {
                const thisPath = [index, defKey, headerCollision];
                const thisDocumentJson = documentNamesToJson[documentName];
                const thisLocation = getPathDocumentCoordinatesPseudoJson(thisPath, thisDocumentJson);
                parseFailures.push(
                    new SchemaParseFailure(
                        otherDocumentName,
                        [otherIndex, otherDefKey, headerCollision],
                        'PathCollision',
                        { document: documentName, path: thisPath, location: thisLocation },
                    ),
                );
            }

            const resHeaderDef = def['->'] as Record<string, object>;
            const otherResHeaderDef = otherDef['->'] as Record<string, object>;

            const resHeaderCollisions = Object.keys(resHeaderDef).filter((k) => k in otherResHeaderDef);
            for (const resHeaderCollision of resHeaderCollisions) {
                const thisPath = [index, '->', resHeaderCollision];
                const thisDocumentJson = documentNamesToJson[documentName];
                const thisLocation = getPathDocumentCoordinatesPseudoJson(thisPath, thisDocumentJson);
                parseFailures.push(
                    new SchemaParseFailure(otherDocumentName, [otherIndex, '->', resHeaderCollision], 'PathCollision', {
                        document: documentName,
                        path: thisPath,
                        location: thisLocation,
                    }),
                );
            }
        }
    }

    if (parseFailures.length > 0) {
        throw new TelepactSchemaParseError(parseFailures, documentNamesToJson);
    }
}

File: lib/go/internal/schema/CatchHeaderCollisions.go
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

package schema

import "sort"

// CatchHeaderCollisions reports duplicate header definitions across Telepact schema documents.
func CatchHeaderCollisions(
	telepactSchemaNameToPseudoJSON map[string][]any,
	headerKeys map[string]struct{},
	keysToIndex map[string]int,
	schemaKeysToDocumentNames map[string]string,
	documentNamesToJSON map[string]string,
) ([]*SchemaParseFailure, error) {
	if len(headerKeys) == 0 {
		return nil, nil
	}

	headers := make([]string, 0, len(headerKeys))
	for key := range headerKeys {
		headers = append(headers, key)
	}

	sort.Slice(headers, func(i, j int) bool {
		docI := schemaKeysToDocumentNames[headers[i]]
		docJ := schemaKeysToDocumentNames[headers[j]]
		if docI == docJ {
			return keysToIndex[headers[i]] < keysToIndex[headers[j]]
		}
		return docI < docJ
	})

	parseFailures := make([]*SchemaParseFailure, 0)

	for i := 0; i < len(headers); i++ {
		for j := i + 1; j < len(headers); j++ {
			defKey := headers[i]
			otherDefKey := headers[j]

			index := keysToIndex[defKey]
			otherIndex := keysToIndex[otherDefKey]

			documentName := schemaKeysToDocumentNames[defKey]
			otherDocumentName := schemaKeysToDocumentNames[otherDefKey]

			pseudoList := telepactSchemaNameToPseudoJSON[documentName]
			otherPseudoList := telepactSchemaNameToPseudoJSON[otherDocumentName]
			if index < 0 || index >= len(pseudoList) || otherIndex < 0 || otherIndex >= len(otherPseudoList) {
				continue
			}

			defMap, ok := pseudoList[index].(map[string]any)
			if !ok {
				continue
			}
			otherDefMap, ok := otherPseudoList[otherIndex].(map[string]any)
			if !ok {
				continue
			}

			headerDef, ok := defMap[defKey].(map[string]any)
			if !ok {
				continue
			}
			otherHeaderDef, ok := otherDefMap[otherDefKey].(map[string]any)
			if !ok {
				continue
			}

			headerCollisions := intersectKeys(headerDef, otherHeaderDef)
			for _, collision := range headerCollisions {
				thisPath := []any{index, defKey, collision}
				location := ResolveDocumentCoordinates(thisPath, documentName, documentNamesToJSON)
				parseFailures = append(parseFailures, NewSchemaParseFailure(
					otherDocumentName,
					[]any{otherIndex, otherDefKey, collision},
					"PathCollision",
					map[string]any{
						"document": documentName,
						"path":     thisPath,
						"location": location,
					},
				))
			}

			resHeaderDef, ok := defMap["->"].(map[string]any)
			if !ok {
				continue
			}
			otherResHeaderDef, ok := otherDefMap["->"].(map[string]any)
			if !ok {
				continue
			}

			resCollisions := intersectKeys(resHeaderDef, otherResHeaderDef)
			for _, collision := range resCollisions {
				thisPath := []any{index, "->", collision}
				location := ResolveDocumentCoordinates(thisPath, documentName, documentNamesToJSON)
				parseFailures = append(parseFailures, NewSchemaParseFailure(
					otherDocumentName,
					[]any{otherIndex, "->", collision},
					"PathCollision",
					map[string]any{
						"document": documentName,
						"path":     thisPath,
						"location": location,
					},
				))
			}
		}
	}

	return parseFailures, nil
}

func intersectKeys(a, b map[string]any) []string {
	result := make([]string, 0)
	for key := range a {
		if _, exists := b[key]; exists {
			result = append(result, key)
		}
	}
	return result
}
